### PR TITLE
Custom TypeScript Types

### PR DESCRIPTION
**Pull request by thesunny**
Fix TypeScript.

This PR improves typing in Slate significantly. It allows developers to extend types like Node within Slate.

See https://github.com/ianstormtaylor/slate/issues/3725

### CODE REVIEW

**Comment by thesunny** [comment]
@mdmjg @BrentFarese 

My intention is to merge this PR. Do you think one of you two can fix the `yarn.lock` issues?

Although I did work on this PR some, I feel a little uncomfortable as you two (and maybe others) are more familiar with the PR as a whole.

Once the conflicts are gone, I'll merge after that.

**Comment by G-Rath** [comment]
@thesunny yarn has automatic conflict resolving, so you should be fine to rebase and just run `yarn install` each time :)

**Comment by BrentFarese** [comment]
cc @timbuckley can you take a look and @thesunny can get this merged?  Thanks!

**Comment by thesunny** [comment]
@BrentFarese @timbuckley

I merged with master and fixed one issue; however, there seems to be one issue failed where it says "deploy/netlify -- Deploy preview failed.". https://app.netlify.com/sites/slatejs/deploys/5f6e3bacf4aeff00072f43a3

I am happy to merge as soon as that issue gets resolved. My apologies as I don't have time to dive into it. It's unclear if that's quick or time-consuming to resolve.

**Comment by BrentFarese** [comment]
> @BrentFarese @timbuckley
> 
> I merged with master and fixed one issue; however, there seems to be one issue failed where it says "deploy/netlify -- Deploy preview failed.". https://app.netlify.com/sites/slatejs/deploys/5f6e3bacf4aeff00072f43a3
> 
> I am happy to merge as soon as that issue gets resolved. My apologies as I don't have time to dive into it. It's unclear if that's quick or time-consuming to resolve.

I don't have too much build experience, but the error seems to be `Error: Cannot find module '@babel/helper-call-delegate'.  It seems like a build issue.  @CameronAckermanSEL do you know what the cause is here?  It might be a short/quick fix.  Thx.

**Comment by BrentFarese** [comment]
FYI this should close #3744 when it's merged too as it fixes the type mentioned in that issue.

**Comment by lcswillems** [comment]
It seems to be related with Netlify cache: https://community.netlify.com/t/error-babel-opt-build-repo-src-index-js-cannot-find-module-babel-helper-call-delegate/11160/6 or https://community.netlify.com/t/error-babel-opt-build-repo-src-index-tsx-cannot-find-module-babel-helper-call-delegate/17456/3

**Comment by BrentFarese** [comment]
> It seems to be related with Netlify cache:
> https://community.netlify.com/t/error-babel-opt-build-repo-src-index-js-cannot-find-module-babel-helper-call-delegate/11160/6
> https://community.netlify.com/t/error-babel-opt-build-repo-src-index-tsx-cannot-find-module-babel-helper-call-delegate/17456/3
> https://community.netlify.com/t/site-deploy-fails-unless-netlify-s-build-cache-is-cleared/12417

Based on these threads, it appears this issue only happens in branch deploys.  And I just viewed the `yarn.lock` changes, which appeared to come in just by merging Master into this branch.  @thesunny or @lcswillems, based on this should we merge to Master and the issue will just clear up since it should only happen on branch deploys?

**Comment by lcswillems** [comment]
@BrentFarese Honestly, I don't know. I never used Netlify...

File: packages/slate/src/interfaces/node.ts [hunk 1]
```diff
@@ -1,14 +1,93 @@
...
+import { ExtendedType } from './custom-types'
 
 /**
  * The `Node` union type represents all of the different types of nodes that
  * occur in a Slate document tree.
  */
 
-export type Node = Editor | Element | Text
+export type BaseNode = Editor | Element | Text
+export type Node = ExtendedType<'Node', BaseNode>
```
Comment (ianstormtaylor) [inline packages/slate/src/interfaces/node.ts:12]: What does this `ExtendedType` helper actually do?

File: packages/slate/src/interfaces/node.ts [hunk 2]
```diff
@@ -1,14 +1,93 @@
...
+    options?: {
+      from?: Path
+      to?: Path
+      reverse?: boolean
+      pass?: (node: NodeEntry) => boolean
+    }
+  ) => Generator<NodeEntry<Text>, void, undefined>
+}
+
+export const Node: NodeInterface = {
```
Comment (ianstormtaylor) [inline packages/slate/src/interfaces/node.ts:90]: Why the need to separate (and duplicate) these interfaces from the definitions?

File: packages/slate/src/interfaces/node.ts (hunk 1 above)
Comment (BrentFarese) [reply packages/slate/src/interfaces/node.ts:12]: It facilitates [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) in Slate so Slate users can specify their own types for `Element`, `Text`, etc. so long as the type extends Slate's base types for those nodes.  Pretty useful so we can get TS compile errors if nodes don't meet a specific type structure defined by the user.

File: packages/slate/src/interfaces/node.ts (hunk 2 above)
Comment (BrentFarese) [reply packages/slate/src/interfaces/node.ts:90]: TS was not inferring the type of interface and transform objects correctly for some reason (ugh), which was causing compile errors.  Extracting the type of the interface/transform object out fixed the issue, but agree it is not necessarily idea/it's duplicative.  We should see in the future if we can remove perhaps for a more clever solution...

**Comment by lcswillems** [comment]
I didn't find the merge in the branch shown on the repo homepage: https://github.com/ianstormtaylor/slate

It is because there are 2 branches: main and master. This is confusing. More over it seems some PR had been merged to master and others to main.

**Comment by BrentFarese** [comment]
Thanks for raising this @lcswillems!  It is a known issue but there wasn't actually a GitHub issue addressing it.  I added an [issue](https://github.com/ianstormtaylor/slate/issues/3996).  We'll take care of moving everything that points to Master to Main as soon as some maintainers can spare some time.

File: packages/slate-react/src/components/editable.tsx [hunk 3]
```diff
@@ -766,7 +767,7 @@ export const Editable = (props: EditableProps) => {
               if (Hotkeys.isRedo(nativeEvent)) {
                 event.preventDefault()
 
-                if (typeof editor.redo === 'function') {
+                if (HistoryEditor.isHistoryEditor(editor)) {
```
Comment (TheSpyder) [inline packages/slate-react/src/components/editable.tsx:770]: This change means `history-editor.ts` from `slate-history` is now included within the `slate-react` bundle.

**Pull request merged**